ntirpc/rpc/gss_internal.h

```diff
@@ -113,30 +112,17 @@ svc_rpc_gss_data *alloc_svc_rpc_gss_data(void)
 }
 
 static inline void
-unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd,
-		       uint32_t flags)
+unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd)
 {
-	u_int refcnt;
-	bool gd_locked = flags & SVC_RPC_GSS_FLAG_LOCKED;
-
-	refcnt = atomic_dec_uint32_t(&gd->refcnt);
+	mutex_lock(&gd->lock);
```

Comment:
As long as we're not done with tearing down the whole thing, 
the relationship between req - auth - gd is intact and can be referenced.
No refcount value is a protection against the usage of a data structure, that can only be achieved by a lock.
The fact that the refcount is zero after doing atomic_dec is only the part of the story that the structure CAN be taken down.
No code path is checking for a refcount > 0 before accessing, we're not having any accessor-methods which could introduce such behaviour.
Anyways, maybe all this cannot happen with the current code but I cannot overlook that for sure now and I don't see any advantage to remove the lock protection from here.
Regarding taking a lock and returning it somewhere else, there's nothing wrong with that and we see that often enough throughout our current code.